contrib/src/main/scala/akka/stream/contrib/Valve.scala

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.stream.contrib
+
+import akka.NotUsed
+import akka.stream.stage.{GraphStageLogic, GraphStageWithMaterializedValue, InHandler, OutHandler}
+import akka.stream._
+import akka.stream.contrib.SwitchMode.{Close, Open}
+
+import scala.concurrent.{Future, Promise}
+
+/**
+  * Pause/ Resume a Flow
+  */
+trait ValveSwitch {
+  /**
+    * Change the state of the valve
+    *
+    * @param mode expected mode to switch on
+    * @return A future that completes with true if the mode did change and false if it already was in the requested mode
+    */
+  def flip(mode: SwitchMode): Future[Boolean]
+}
+
+
+object Valve {
+  /**
+    * Factory for [[Valve]] instances.
+    *
+    * @tparam A input and output type
+    * @return [[Valve]] Opened instance
+    *
+    */
+  def apply[A](): Valve[A] = new Valve[A](SwitchMode.Open)
+
+  /**
+    * Factory for [[Valve]] instances.
+    *
+    * @tparam A input and output type
+    * @return [[Valve]] instance
+    *
+    */
+  def apply[A](mode: SwitchMode): Valve[A] = new Valve[A](mode)
+
+  /**
+    * Java API: Factory for [[Valve]] instances.
+    *
+    * @tparam A input and output type
+    * @return [[Valve]] Opened instance
+    */
+  def create[A](): Valve[A] = new Valve[A](SwitchMode.Open)
+
+  /**
+    * Java API: Factory for [[Valve]] instances.
+    *
+    * @tparam A input and output type
+    * @return [[Valve]] instance
+    */
+  def create[A](mode: SwitchMode): Valve[A] = new Valve[A](mode)
+
+}
+
+/**
+  * Materializes into a [[ValveSwitch]] which provides a the method flip that stops or restarts the flow of elements passing through the stage. As long as the valve is closed it will backpressure.
+  *
+  * Note that closing the valve could result in one element being buffered inside the stage, and if the stream completes or fails while being closed, that element may be lost.
+  *
+  * @param mode : state of the valve at the startup of the flow (by default Open)
+  */
+class Valve[A](mode: SwitchMode) extends GraphStageWithMaterializedValue[FlowShape[A, A], ValveSwitch] {
+
+  private val in: Inlet[A] = Inlet[A]("valve.in")
+  private val out: Outlet[A] = Outlet[A]("valve.out")
```

Comment:
Don't make them private, being public allows them to be used for wiring with other stages in the GraphDSL